airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py

```diff
@@ -1476,6 +1476,20 @@ def _merge_stream_slicers(
                     stream_cursor=cursor_component,
                 )
         elif model.incremental_sync:
+            if model.retriever.type == "AsyncRetriever":
+                if model.incremental_sync.type != "DatetimeBasedCursor":
+                    # We are currently in a transition to the Concurrent CDK and AsyncRetriever can only work with the support or unordered slices (for example, when we trigger reports for January and February, the report in February can be completed first). Once we have support for custom concurrent cursor or have a new implementation available in the CDK, we can enable more cursors here.
+                    raise ValueError("AsyncRetriever with cursor other than DatetimeBasedCursor is not supported yet")
+                if model.retriever.partition_router:
+                    # Note that this development is also done in parallel to the per partition development which once merged we could support here by calling `create_concurrent_cursor_from_perpartition_cursor`
+                    raise ValueError("Per partition state is not supported yet for AsyncRetriever")
+                return self.create_concurrent_cursor_from_datetime_based_cursor(
```

Comment:
This is one part of the transition to `ConcurrentCursor` that feels a bit jank because we're passing back a `ConcurrentCursor` to persist to the various low-code components and that is no longer technically the same `StreamSlicer` (low-code) interface that was to be expected. So at least in the short term our typing is going to be incorrect.

I think I can live with this since we do ultimately want to drop these low-code interfaces, but I'll add a comment here and we'll need to override mypy if we agree that this is tolerable